### ISSUE TITLE
Port async TestCase to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6033]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6033|
|Type|enhancement|
|Created|2012-10-01 14:06:29Z|
|Branch|https://github.com/twisted/twisted/tree/testcase-py3-6033-2|

`twisted.trial.unittest.TestCase` should run on Python 3.

<details><summary>Searchable metadata</summary>

```
trac-id__6033 6033
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_testcase_py3_6033_2 branches/testcase-py3-6033-2
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1349100389000000 1349100389000000
changetime__1349359177000000 1349359177000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [35867]) Branching to 'testcase-py3-6033'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst set owner to @exarkun|
|-|-|

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/testcase-py3-6033

In order to get decent coverage I had to port an unfortunately large bunch of modules, and make `reporter` at least import (it's used by tests, not by `TestCase` at least). On the plus side there weren't many changes to make in any of them.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @itamarst|
|-|-|

Thanks.

  1. At the end of `twisted/trial/test/test_deferred.py`, the `_PY3` conditional doesn't seem like it needs to be conditional.  trial actually loads `TestTester` too, but doesn't run it because it has no `test_` methods.  Whereas the stdlib runner seems to believe that `runTest` is a test (and this manifests as a problem elsewhere too, so potentially we should fix it, but I don't think it's related to the Python 3 port).  So anyway, just delete `TestTester` unconditionally, I think.  Also, the tests in this module are wacko. :/  It seems like they should be written differently (new ticket):
    1. `runTest` is part of the `TestCase` API, it's a bad idea to re-use it for something else.
    1. Other trial tests use the loader to get something runnable instead of doing a manual loading step like `getTest` is doing.
    1. Some of these tests are slow, waiting for a 0.1s timeout before proceeding.
    1. No docs anywhere :/
  1. Looks like `test_threads` got lost.
  1. in `twisted/trial/test/test_tests.py`, a skip message is duplicated many times.  The idiom of defining the skip message once at the beginning of a module and then re-using it is a sort of nice one I prefer (eg how `ipv6Skip` works in `twisted/internet/test/test_tcp.py`).
  1. It would be better to port the `twisted.python.deprecate` dependency first instead of poking at a couple of its edges in this branch without being able to run its tests.  Without running its tests, how do we know the changes are correct?
  1. in `twisted/trial/reporter.py`
    1. under what conditions does the `collections.OrderedDict` import fail?  `OrderedDict` appears to be in Python 2.7.
    1. why break the link to `untilConcludes`?  That seems bad.
  1. in `twisted/trial/test/test_reporter.py`, just ugh ugh ugh ugh

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

See [35887] for fixes mentioned below.

  1. Made the delete unconditional, opened [#6038](https://github.com/twisted/twisted/issues/6038).
  1. Added it back.
  1. Unified them all into a single message constant. 
  1. Will do.
  1. `OrderedDict` isn't present in Python 2.6. The link to `untilConcludes` was causing pydoctor errors, probably because of that bug mwhudson fixed recently.
  1. It was like that when I got here.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

OK, [#6036](https://github.com/twisted/twisted/issues/6036) is up for review.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

(In [35928]) Branching to 'testcase-py3-6033-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst set owner to @exarkun|
|-|-|

OK, merged forward. New test run:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/testcase-py3-6033-2

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @itamarst|
|-|-|

Thanks

  1. `twisted/trial/_asynctest.py`
    1. zope.interface import belongs above the Twisted imports
    1. The `twisted.internet.reactor` import hidden in `TestCase._run` can be moved to module-scope now?  Perhaps we also need a ticket for turning it into a parameter... somewhere (on the results object?  as an argument to a new `run`-like method that the runner can call instead of `run`?)
    1. The timeout support in `TestCase._run` should probably be based on `Deferred` cancellation, I suppose?  Hard to say, since there is still no real cancellation documentation.  Still, I suspect it should be.  Can you file a ticket for that?
    1. Another nested reactor import in `_runFixturesAndTest`.
    1. And another in `_wait`
  1. `twisted/trial/reporter.py`
    1. The `untilConcludes` link is still being broken.  Maybe you said something about that on IRC but I forget what.  It doesn't seem like we need to break this link though.
    1. zope.interface import here should also be moved above Twisted imports
  1. `twisted/trial/test/test_asyncassertions.py`
    1. The stdlib unittest import belongs above the Twisted imports
    1. Many tests in this module fail to follow the coding standard.  Can you file a ticket for fixing that?
  1. `twisted/trial/test/test_deferred.py`
    1. stdlib unittest import should be above Twisted imports
    1. Many tests in this module fail to follow the coding standard.  Can you file a ticket for fixing that?
    1. I guess there should be a ticket for the `del TestLoader` thing.  We shouldn't really need to be doing that.
  1. `twisted/trial/test/test_tests.py`
    1. stdlib unittest import belongs above the Twisted imports
    1. There are references to [[#6034](https://github.com/twisted/twisted/issues/6034)](https://github.com/twisted/twisted/issues/6034) in the code.  #6034 was closed as a duplicate of [#5964](https://github.com/twisted/twisted/issues/5964).
    1. Many of the tests in this module do not follow the coding standard.  Can you file a ticket for fixing this?

Overall, seems good.  Merge when the above are addressed.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst commented|
|-|-|

  1. [#6047](https://github.com/twisted/twisted/issues/6047) and [#6048](https://github.com/twisted/twisted/issues/6048).
  1. I'm pretty sure the link is broken due to the bug Michael Hudson recently fixed in pydoctor. I'll try another way, which will definitely work if not now then eventually.
  1. [#6049](https://github.com/twisted/twisted/issues/6049).
  1. [#6050](https://github.com/twisted/twisted/issues/6050), expanded [#6038](https://github.com/twisted/twisted/issues/6038) for the del thing.
  1. [#6051](https://github.com/twisted/twisted/issues/6051).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst commented|
|-|-|

(In [35930]) Address review comments. Refs [#6033](https://github.com/twisted/twisted/issues/6033)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_12"></a>|@itamarst commented|
|-|-|

Oh. There's a reason the reactor isn't imported at module level - it prevents trial from choosing a reactor:

```
./bin/trial: The specified reactor cannot be used, failed with error: reactor already installed.
See the list of available reactors with --help-reactors
```

So I will revert that particular change and add a comment explaining why there's no import at top level.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst set status to `closed`|
|-|-|

(In [35932]) Merge testcase-py3-6033-2.

Author: itamar
Review: exarkun
Fixes: [#6033](https://github.com/twisted/twisted/issues/6033)

twisted.trial.unittest.TestCase now runs on Python 3.
